src/stakeinput.cpp

```diff
@@ -18,81 +19,69 @@ CZPivStake::CZPivStake(const libzerocoin::CoinSpend& spend)
-    return GetChecksumHeight(nChecksum, denom);
-}
-
-int CZPivStake::GetChecksumHeightFromSpend()
-{
-    return GetChecksumHeight(nChecksum, denom);
-}
-
 uint32_t CZPivStake::GetChecksum()
 {
     return nChecksum;
 }
 
-// The zPIV block index is the first appearance of the accumulator checksum that was used in the spend
-// note that this also means when staking that this checksum should be from a block that is beyond 60 minutes old and
-// 100 blocks deep.
+/*
+ * LEGACY: Kept for IBD in order to verify zerocoin stakes occurred when zPoS was active
+ * Find the first occurrence of a certain accumulator checksum.
+ * Return block index pointer or nullptr if not found
+ */
+
 CBlockIndex* CZPivStake::GetIndexFrom()
 {
-    if (pindexFrom)
-        return pindexFrom;
-
+    // First look in the legacy database
     int nHeightChecksum = 0;
-
-    if (fMint)
-        nHeightChecksum = GetChecksumHeightFromMint();
-    else
-        nHeightChecksum = GetChecksumHeightFromSpend();
-
-    if (nHeightChecksum < Params().Zerocoin_StartHeight() || nHeightChecksum > chainActive.Height()) {
-        pindexFrom = nullptr;
-    } else {
-        //note that this will be a nullptr if the height DNE
-        pindexFrom = chainActive[nHeightChecksum];
+    if (zerocoinDB->ReadAccChecksum(nChecksum, denom, nHeightChecksum)) {
+        return chainActive[nHeightChecksum];
     }
 
-    return pindexFrom;
+    // Not found. Scan the chain.
+    CBlockIndex* pindex = chainActive[Params().Zerocoin_StartHeight()];
+    if (!pindex) return nullptr;
+    const int last_block = Params().Zerocoin_Block_Last_Checkpoint();
+    while (pindex && pindex->nHeight <= last_block) {
+        if (ParseAccChecksum(pindex->nAccumulatorCheckpoint, denom) == nChecksum) {
+            // Found. Save to database and return
+            zerocoinDB->WriteAccChecksum(nChecksum, denom, pindex->nHeight);
+            return pindex;
+        }
+        //Skip forward in groups of 10 blocks since checkpoints only change every 10 blocks
+        if (pindex->nHeight % 10 == 0) {
+            pindex = chainActive[pindex->nHeight + 10];
+            continue;
+        }
+        pindex = chainActive.Next(pindex);
+    }
+    return nullptr;
 }
 
 CAmount CZPivStake::GetValue()
 {
     return denom * COIN;
 }
 
-//Use the first accumulator checkpoint that occurs 60 minutes after the block being staked from
-// In case of regtest, next accumulator of 60 blocks after the block being staked from
+// Use the first accumulator checkpoint that occurs 60 minutes after the block being staked from
 bool CZPivStake::GetModifier(uint64_t& nStakeModifier)
 {
     CBlockIndex* pindex = GetIndexFrom();
     if (!pindex)
         return error("%s: failed to get index from", __func__);
 
     if(Params().IsRegTestNet()) {
-        // Stake modifier is fixed for now, move it to 60 blocks after this pindex in the future..
-        nStakeModifier = pindexFrom->nStakeModifier;
+        nStakeModifier = 0;
```

Comment:
Know that this is not from this PR but why we are not using the regular flow for this?

The current modifier flow is much straightforward than the previous one.

Reply:
Yep, sure.
I've just opened #1309 to keep track of this.